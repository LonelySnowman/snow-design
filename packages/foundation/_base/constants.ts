export const BASE_CLASS_PREFIX = 'snow';
