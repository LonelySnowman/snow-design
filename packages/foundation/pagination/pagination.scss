@import "./variables.scss";

$module: #{$prefix}-pagination;

.#{$module} {
  display: flex;
  list-style: none;
  padding: $spacing-pagination_item-margin;
  align-items: center;
  margin-block-start: 0;
  margin-block-end: 0;

  &-disabled {
    cursor: not-allowed;

    .#{$module}-total {
      color: $color-pagination_item-text-disabled;
    }
  }

  &-item {
    @include font-size-regular;
    display: flex;
    justify-content: center;
    align-items: center;
    height: $size-pagination_item;
    min-width: $size-pagination_item;
    margin-inline-end: $spacing-pagination_item_margin;
    border-radius: var(--snow-border-radius);
    cursor: pointer;
    box-sizing: border-box;
    user-select: none;

    &:hover {
      background-color: $color-pagination_item-bg-hover;
    }

    &-active {
      border: 1px solid $color-pagination_item-border-active;
      color: $color-pagination_item-active;

      &:hover {
        border-color: $color-pagination_item-border-active-hover;
        color: $color-pagination_item-active-hover;
        background-color: transparent;
      }
    }

    &-disabled {
      color: $color-pagination_item-text-disabled;
      cursor: not-allowed;
      &:hover {
        background-color: transparent;
      }
    }
  }

  &-total {
    @include font-size-regular;
    margin-right: $spacing-pagination_item-margin;
  }

  &-prev,
  &-next {
    padding: 4px;
    cursor: pointer;
    background-color: transparent;
    color: $color-pagination_item-active;
    &:hover {
      background-color: transparent;
    }
    &.#{$module}-item-disabled {
      cursor: not-allowed;
      color: $color-pagination_item-disabled;
    }
  }
}
