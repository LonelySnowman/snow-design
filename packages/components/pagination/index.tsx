import React, { useEffect, useState } from 'react';
import '@snow-design/foundation/pagination/pagination.scss'
import { cssClasses } from '@snow-design/foundation/pagination/constants';
import classNames from 'classnames';
import { useLocale } from '../_locale';
import usePaginationFoundation from "@snow-design/foundation/pagination/foundation";
import useMergedState from "../_utils/hooks/useMergedState";
import { CssProps } from "@snow-design/components/_types";

const prefixCls = cssClasses.PREFIX;

export interface PaginationProps extends CssProps {
    pageSize: number;
    total: number;
    currentPage?: number;
    defaultCurrentPage?: number;
    showTotal?: boolean;
    onChange?: (page: number, pageSize: number) => void;
    disabled?: boolean;
    nextText?: React.ReactNode;
    prevText?: React.ReactNode;
}

const Button: React.FC<PaginationProps> = (props) => {
    const {
        currentPage,
        defaultCurrentPage = 1,
        pageSize,
        total,
        showTotal = false,
        disabled = false,
        onChange,
        nextText,
        prevText,
        className,
        style,
    } = props;

    if (currentPage && !onChange) {
        console.warn(
            'Warning: you have provide currentPage prop for pagination but without onChange handler ,' +
            ' this will cause no-change when you change page. '
        );
    }

    const [btnDisabled, setBtnDisabled] = useState<{prevIsDisabled: boolean, nextDisabled: boolean}>({
        prevIsDisabled: false,
        nextDisabled: false
    })
    const [pageList, setPageList] = useState([]);
    const [curPageValue, setCurPageValue] = useMergedState<number>(defaultCurrentPage, {
        value: currentPage,
        onChange: (value) => {
            foundation.goPage(value);
        }
    });

    const [contextLocale] = useLocale('Pagination');
    const isControlComponent = currentPage !== undefined;
    const foundation = usePaginationFoundation({
        getProps() {
            return props;
        },
        getStates() {
            return {
                currentPage: curPageValue,
                total,
                pageSize
            }
        },
        setCurrentPage(pageIndex: number) {
            setCurPageValue(pageIndex);
        },
        setPageList(pageList){
            setPageList(pageList);
        },
        setDisabled: (prevIsDisabled: boolean, nextIsDisabled: boolean) => {
            setBtnDisabled({ prevIsDisabled: prevIsDisabled, nextDisabled: nextIsDisabled });
        },
        notifyChange: (pageIndex: number, pageSize: number) => {
            if (typeof onChange === "function") {
                onChange(pageIndex, pageSize);
            }
        }
    })

    useEffect(() => {
        foundation.init();
    }, []);

    const renderPrevBtn = () => {
        const isDisabled = btnDisabled.prevIsDisabled || disabled;
        const preClassName = classNames({
            [`${prefixCls}-item`]: true,
            [`${prefixCls}-prev`]: true,
            [`${prefixCls}-item-disabled`]: isDisabled,
        });
        return (
          <li
            role="button"
            aria-disabled={isDisabled}
            aria-label="Previous"
            onClick={() => !isDisabled && foundation.goPrev()}
            className={preClassName}
          >
              {prevText || contextLocale.previous}
          </li>
        );
    }

    const renderNextBtn = () => {
        const isDisabled = btnDisabled.nextDisabled || disabled;
        const nextClassName = classNames({
            [`${prefixCls}-item`]: true,
            [`${prefixCls}-next`]: true,
            [`${prefixCls}-item-disabled`]: isDisabled,
        });
        return (
          <li
            role="button"
            aria-disabled={isDisabled}
            aria-label="Next"
            onClick={() => !isDisabled && foundation.goNext()}
            className={nextClassName}
          >
              {/* @todo: Icon 模块待设计 */}
              {nextText || contextLocale.next}
          </li>
        );
    }

    const renderPageList = () => {
        return pageList.map((page, i) => {
            const pageListClassName = classNames(`${prefixCls}-item`, {
                [`${prefixCls}-item-active`]: curPageValue === page,
                [`${prefixCls}-item-all-disabled`]: disabled,
                [`${prefixCls}-item-all-disabled-active`]: curPageValue === page && disabled,
            });
            const pageEl = (
              <li
                key={`${page}${i}`}
                onClick={() => !disabled && foundation.goPage(page, !isControlComponent)}
                className={pageListClassName}
                aria-label={page === '...' ? 'More' : `Page ${page}`}
                aria-current={curPageValue === page ? "page" : false}
              >
                  {page}
              </li>
            );
            return pageEl;
        });
    }

    const paginationCls = classNames(className, `${prefixCls}`, { [`${prefixCls}-disabled`]: disabled });

    const totalNum = Math.ceil(total / pageSize);
    const totalToken = contextLocale.total.replace('${total}', totalNum.toString());

    return (
      <ul className={paginationCls} style={style}>
          {showTotal ? (
            <span className={`${prefixCls}-total`}>{totalToken}</span>
          ) : null}
          {renderPrevBtn()}
          {renderPageList()}
          {renderNextBtn()}
      </ul>
    );
}

export default Button;
