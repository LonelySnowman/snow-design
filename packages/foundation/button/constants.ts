import { BASE_CLASS_PREFIX } from '../_base/constants';

export const cssClasses = {
    PREFIX: `${BASE_CLASS_PREFIX}-button`,
};
