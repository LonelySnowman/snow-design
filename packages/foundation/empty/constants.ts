import { BASE_CLASS_PREFIX } from '../_base/constants';

const cssClasses = {
    PREFIX: `${BASE_CLASS_PREFIX}-empty`
};

export {
    cssClasses,
};
