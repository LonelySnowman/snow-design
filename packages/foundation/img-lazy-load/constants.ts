import { BASE_CLASS_PREFIX } from '../_base/constants';

const cssClasses = {
    PREFIX: `${BASE_CLASS_PREFIX}-img-lazy-load`,
};

export { cssClasses };
