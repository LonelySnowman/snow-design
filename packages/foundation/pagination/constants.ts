import { BASE_CLASS_PREFIX } from '../_base/constants';

const cssClasses = {
    PREFIX: `${BASE_CLASS_PREFIX}-pagination`
};

const strings = {};

const numbers = {
    PAGE_SHOW_MAX: 7,
    REST_PAGE_SHOW_MAX: 5,
    DEFAULT_PAGE_SIZE: 10,
    PAGE_SIZE_OPTION: [10, 20, 40, 100],
    REST_PAGE_MAX_SIZE: 1000000,
}

export { cssClasses, strings, numbers };
