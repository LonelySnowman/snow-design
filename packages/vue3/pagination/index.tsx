import { defineComponent, onMounted, reactive, CSSProperties } from 'vue';
// import '@snow-design/foundation/pagination/pagination.scss'
import { cssClasses } from '@snow-design/foundation/pagination/constants';
import classNames from 'classnames';
import usePaginationFoundation from "@snow-design/foundation/pagination/foundation";
import { anyType, objectType, withInstall, VueNode } from '../_utils/type';
import useLocale from "../_locale/useLocale";

const prefixCls = cssClasses.PREFIX;

export const paginationProps = () => ({
  currentPage: Number,
  defaultCurrentPage: Number,
  pageSize: Number,
  total: Number,
  showTotal: Boolean,
  onChange: Function,
  disabled: Boolean,
  nextText: anyType<VueNode>(),
  prevText: anyType<VueNode>(),
  class: String,
  style: objectType<CSSProperties>(),
});

const Pagination = defineComponent({
  name: 'SPagination',
  compatConfig: { MODE: 3 },
  inheritAttrs: false,
  props: paginationProps(),
  setup(props: any) {
    const {
      currentPage,
      defaultCurrentPage = 1,
      pageSize,
      total,
      showTotal,
      disabled = false,
      onChange,
      nextText,
      prevText,
      className,
      style,
    } = props;

    console.log(props.pageSize);

    const [contextLocale] = useLocale('Pagination');

    const paginationState = reactive({
      prevIsDisabled: false,
      nextIsDisabled: false,
      pageList: [],
      curPageValue: defaultCurrentPage,
    })

    const foundation = usePaginationFoundation({
      // LJQFLAG 抽离公共 Foundation
      getProps() {
        return props;
      },
      getStates() {
        return {
          currentPage: paginationState.curPageValue,
          total,
          pageSize
        }
      },
      setCurrentPage(pageIndex: number) {
        paginationState.curPageValue = pageIndex;
      },
      setPageList(pageList){
        paginationState.pageList = pageList;
      },
      setDisabled: (prevIsDisabled: boolean, nextIsDisabled: boolean) => {
        paginationState.prevIsDisabled = prevIsDisabled;
        paginationState.nextIsDisabled = nextIsDisabled;
      },
      notifyChange: (pageIndex: number, pageSize: number) => {
        onChange(pageIndex, pageSize);
      }
    })

    onMounted(() => {
      foundation.init();
    });

    const renderPrevBtn = () => {
      const isDisabled = paginationState.prevIsDisabled || disabled;
      const preClassName = classNames({
        [`${prefixCls}-item`]: true,
        [`${prefixCls}-prev`]: true,
        [`${prefixCls}-item-disabled`]: isDisabled,
      });
      return (
        <li
          role="button"
          aria-disabled={isDisabled}
          aria-label="Previous"
          onClick={e => !isDisabled && foundation.goPrev()}
          class={preClassName}
        >
          {/* LJQFLAG 自定义 ICON */}
          {/*{prevText || <IconChevronLeft size="large" />}*/}
          {prevText || contextLocale.previous}
        </li>
      );
    }

    const renderNextBtn = () => {
      const isDisabled = paginationState.nextIsDisabled || disabled;
      const nextClassName = classNames({
        [`${prefixCls}-item`]: true,
        [`${prefixCls}-next`]: true,
        [`${prefixCls}-item-disabled`]: isDisabled,
      });
      return (
        <li
          role="button"
          aria-disabled={isDisabled}
          aria-label="Next"
          onClick={e => !isDisabled && foundation.goNext()}
          class={nextClassName}
        >
          {/* LJQFLAG 自定义 ICON */}
          {/*{prevText || <IconChevronLeft size="large" />}*/}
          {nextText || contextLocale.next}
        </li>
      );
    }

    const renderPageList = () => {
      return paginationState.pageList.map((page, i) => {
        const pageListClassName = classNames(`${prefixCls}-item`, {
          [`${prefixCls}-item-active`]: paginationState.curPageValue === page,
          [`${prefixCls}-item-all-disabled`]: disabled,
          [`${prefixCls}-item-all-disabled-active`]: paginationState.curPageValue === page && disabled,
        });
        const pageEl = (
          <li
            key={`${page}${i}`}
            onClick={() => !disabled && foundation.goPage(page)}
            class={pageListClassName}
            aria-label={page === '...' ? 'More' : `Page ${page}`}
            aria-current={paginationState.curPageValue === page ? "page" : false}
          >
            {page}
          </li>
        );
        return pageEl;
      });
    }

    const paginationCls = classNames(className, `${prefixCls}`, { [`${prefixCls}-disabled`]: disabled });
    const showTotalCls = `${prefixCls}-total`;

    const totalNum = Math.ceil(total / pageSize);
    const totalToken = contextLocale.total.replace('${total}', totalNum.toString());

    return () => (
      <ul class={paginationCls} style={style}>
        {showTotal ? (
          <span class={showTotalCls}>{totalToken}</span>
        ) : null}
        {renderPrevBtn()}
        {renderPageList()}
        {renderNextBtn()}
      </ul>
    )
  }
})

export default withInstall(Pagination);
