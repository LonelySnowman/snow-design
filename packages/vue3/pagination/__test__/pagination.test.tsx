import mountTest from "#test/vue/mountTest";
import styleTest from "#test/vue/styleTest";
import Pagination from '../index';
import { render } from "@testing-library/vue";
import { cssClasses } from '@snow-design/foundation/pagination/constants';
import '@testing-library/jest-dom'

describe('Pagination', () => {
    const prefix = cssClasses.PREFIX;
    mountTest(<Pagination pageSize={10} total={20} />);
    styleTest(Pagination);

    it('renders correctly', () => {
        const { container } = render(<Pagination pageSize={10} total={20} />)
        expect(container.firstChild).toMatchSnapshot();
    })

    it('having correct pagination', () => {
        const { container } = render(<Pagination pageSize={2} total={20} />)
        expect(container.getElementsByTagName('li').length).toEqual(9)
        expect(container.firstChild).toHaveClass(prefix);
    })
});
