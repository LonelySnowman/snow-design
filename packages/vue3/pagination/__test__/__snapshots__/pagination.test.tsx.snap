// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination renders correctly 1`] = `
<ul
  class="snow-pagination"
>
  <!---->
  <li
    aria-disabled="true"
    aria-label="Previous"
    class="snow-pagination-item snow-pagination-prev snow-pagination-item-disabled"
    role="button"
  >
    Previous
  </li>
  
  <li
    aria-current="page"
    aria-label="Page 1"
    class="snow-pagination-item snow-pagination-item-active"
  >
    1
  </li>
  <li
    aria-current="false"
    aria-label="Page 2"
    class="snow-pagination-item"
  >
    2
  </li>
  
  <li
    aria-disabled="false"
    aria-label="Next"
    class="snow-pagination-item snow-pagination-next"
    role="button"
  >
    Next
  </li>
</ul>
`;
